syntax = "proto3";

option objc_class_prefix = "HLW";

// The ClientHandler service definition.
service ClientHandler {
  // Lists users
  rpc ListUsers (ListRequest) returns (ListReply) {}
  // Requests to Login
  rpc Login (LoginRequest) returns (LoginReply) {}
  // Requests to Logout
  rpc Logout (LogoutRequest) returns (LogoutReply) {}
  // Attempts to delete account
  rpc Delete (DeleteRequest) returns (DeleteReply) {}
  // Checks a file
  rpc Check (CheckRequest) returns (CheckReply) {}
  // Attempts to upload a file
  rpc Upload (stream UploadRequest) returns (UploadReply) {}
  // Attempts to sync files
  rpc Sync (SyncRequest) returns (stream SyncReply) {}
}

message CheckRequest {
  string user = 1;
  bytes hash = 2;
  int32 clock = 3;
}

message CheckReply {
  int32 status = 1;
  string errormessage = 2;
  bool sendupdate = 3;
}

message SyncRequest {
  string user = 1;
  repeated Metadata metadata = 2;
}

message SyncReply {
    oneof test_oneof {
        Metadata meta = 2;
        bytes file = 1;
  }
}

message UploadRequest{
  oneof test_oneof {
        Metadata meta = 2;
        bytes file = 1;
  }
}

message Metadata{
  int32 clock = 1;
  string user = 2;
  bytes hash = 3;
  int64 MAC = 4;
  string filename = 5;
  string filepath = 6;
}

message UploadReply {
  int32 status = 1;
  string errormessage = 2;
  bool success = 3;
}

message LoginRequest {
  string user = 1;
}

message LogoutRequest{
    string user = 1;
}

message ListRequest {
    string args = 1;
}

message DeleteRequest {
    string user = 1;
}

message GetRequest {
  string user = 1;
}

message LoginReply {
  int32 status = 1;
  string errormessage = 2;
  string user = 3;
}

message LogoutReply {
    int32 status = 1;
    string errormessage = 2;
    string user = 3;
}

message ListReply {
  int32 status = 1;
  string errormessage = 2;
  string wildcard = 3;
  repeated string user = 4;
}

message DeleteReply {
    int32 status = 1;
    string errormessage = 2;
    string user = 3;
}


message UnreadMessage {
  string sender = 1;
  string message = 2;
  string receiver = 3;
}

message GetReply {
  int32 status = 1;
  string errormessage = 2;
  repeated UnreadMessage message = 3;
}
